Add http_parser::header_uint

pavel-pimenov opened the pull request.

arvidn commented on src/web_peer_connection.cpp:
```diff
@@ -578,8 +578,8 @@ void web_peer_connection::handle_error(int const bytes_left)
 	// pad files will make it complicated
-	int retry_time = atoi(m_parser.header("retry-after").c_str());
-	if (retry_time <= 0) retry_time = m_settings.get_int(settings_pack::urlseed_wait_retry);
+	int const retry_time = m_parser.header_uint("retry-after", 
```
I prefer the comma on the new line, to make it clearer that this line is a continuation of the statement from the previous line.

arvidn commented on include/libtorrent/http_parser.hpp:
```diff
@@ -67,6 +67,14 @@ namespace libtorrent {
 			if (i == m_header.end()) return empty;
 			return i->second;
 		}
+		unsigned header_uint(char const* key, unsigned const def_value) const
```
My gut reaction is that I don't think parsing an integer from a string should be a member function. This should be a free function. I suppose you also provide the default in case the header isn't found, which in a way is more robust than to parse an empty string. Maybe it could just have a different name, like ``header_int()``, and return a ``std::int64_t``. What should it do on overflow? either throw an exception or return the default.

More fundamentally though, this should not return an unsigned integer. Unsigned integers have modulo 2 arithmetic, and that's not appropriate for any value parsed out of an http header (at least I find it unlikely). The default should also not be ``unsigned``.

It should also not take a c-string. It should take a ``string_view``.

arvidn commented:
you may have noticed from the comments associated with this commit that you need to rebase on top of the new master

arvidn commented:
I don't think your rebase was successful. the abandoned commit is still in the history

pavel-pimenov commented:
help git rebase? :(

arvidn commented:
make sure you have the latest master. you will need to reset since there's a discontinuity. I think it would be something along these lines. Note that I'm not 100% confident I got all this right. You may want to make a backup of your repo before running this.

```
git checkout master
git reset origin master
git pull
git reset --hard
```

then checkout your branch and remove the merge commits from it:

```
git checkout add-header_uint
git reset 6f8c52f5c8908bf5bcb10f1d38a7c81aeb48472f
```

Then rebase. This is meant to take all commits from the abandoned master commit to your head, and place them on top of the new master.

```
git rebase --onto master 483e1a1de7822206af2db1254537ad27360407e5 add-header_uint
```

pavel-pimenov commented:
Q:\git\libtorrent-ppa>git checkout master
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.
Q:\git\libtorrent-ppa>git reset origin master
Q:\git\libtorrent-ppa>git pull
Already up-to-date.
Q:\git\libtorrent-ppa>git reset --hard
HEAD is now at 80c4962 Merge branch 'master' of https://github.com/arvidn/libtorrent

Q:\git\libtorrent-ppa>git checkout add-header_uint
Switched to branch 'add-header_uint'
Your branch is up-to-date with 'origin/add-header_uint'.

Q:\git\libtorrent-ppa>git reset 6f8c52f5c8908bf5bcb10f1d38a7c81aeb48472f
Unstaged changes after reset:
M       src/http_seed_connection.cpp
M       src/web_peer_connection.cpp

Q:\git\libtorrent-ppa>git rebase --onto master 483e1a1de7822206af2db1254537ad27360407e5 add-header_uint
Cannot rebase: You have unstaged changes.
Please commit or stash them.

pavel-pimenov commented:
Create new PR https://github.com/arvidn/libtorrent/pull/2472
